components/schema-migrator/migrations/director/20210915134043_add_providers_to_views.up.sql

```diff
@@ -0,0 +1,325 @@
+BEGIN;
+
+CREATE OR REPLACE FUNCTION apps_subaccounts_func()
+    RETURNS TABLE
+            (
+                id                 uuid,
+                tenant_id          uuid,
+                provider_tenant_id uuid
+            )
+    LANGUAGE plpgsql
+AS
+$$
+BEGIN
+    RETURN QUERY
+        SELECT l.app_id                 AS id,
+               asa.selector_value::uuid AS tenant_id,
+               asa.selector_value::uuid AS provider_tenant_id
+        FROM labels l
+                 -- 2) Get subaccounts in those scenarios (Putting a subaccount in a
+                 -- scenario will reflect on creating an ASA for the subaccount.
+                 JOIN automatic_scenario_assignments asa
+                      ON asa.tenant_id = l.tenant_id AND l.value ? asa.scenario::text AND
+                         asa.selector_key::text = 'global_subaccount_id'::text
+             -- 1) Get all scenario labels for applications
+        WHERE l.app_id IS NOT NULL
+          AND l.key::text = 'scenarios'::text;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION consumers_provider_for_runtimes_func()
+    RETURNS TABLE
+            (
+                provider_tenant  jsonb,
+                consumer_tenants jsonb
+            )
+    LANGUAGE plpgsql
+AS
+$$
+BEGIN
+    RETURN QUERY
+        SELECT l1.value -> 0 AS provider_tenant, l2.value AS consumer_tenants
+        FROM (SELECT * FROM labels WHERE key::text = 'global_subaccount_id') l1 -- Get the subaccount for each runtime
+                 JOIN (SELECT * FROM labels WHERE key::text = 'consumer_subaccount_ids') l2 -- Get all the consumer subaccounts for each runtime
+                      ON l1.runtime_id = l2.runtime_id AND l1.runtime_id IS NOT NULL;
+END
+$$;
+
+DROP VIEW IF EXISTS tenants_specifications;
+DROP VIEW IF EXISTS tenants_apis;
+
+CREATE OR REPLACE VIEW tenants_apis
+            (tenant_id, provider_tenant_id, id, app_id, name, description, group_name, default_auth, version_value, version_deprecated,
+             version_deprecated_since, version_for_removal, ord_id, short_description, system_instance_aware,
+             api_protocol, tags, countries, links, api_resource_links, release_status, sunset_date, changelog_entries,
+             labels, package_id, visibility, disabled, part_of_products, line_of_business, industry, ready, created_at,
+             updated_at, deleted_at, error, implementation_standard, custom_implementation_standard,
+             custom_implementation_standard_description, target_urls, extensible, successors, resource_hash)
+AS
+WITH apps_subaccounts AS (SELECT * FROM apps_subaccounts_func()),
+     consumers_provider_for_runtimes AS (SELECT * FROM consumers_provider_for_runtimes_func())
```

Comment:
Are those needed? Can you just Select from the function below?

Reply:
I thought there was a performance benefit since the apps_subaccounts CTE is used more than once so I thought that it would get calculated once and then read multiple times. However, after some reading on the Internet, I came to know that CTEs are not very performant in Postgresql. I executed an `EXPLAIN ANALYZE ...` function and actually got better results with functions only, so yes, I will remove them.